### ISSUE TITLE
In JIRA: Reflect attribute dependencies logic in manifest column ordering

### ISSUE THREAD

**Issue by milen-sage** [Sage-Bionetworks/schematic#214]
Currently columns are ordered based on the schema definition, placing all unconditionally required columns first. However, the order of the conditionally required columns does not reflect their relationship with unconditionally required columns. We can try to improve the column ordering logic to address that. Some thought's needed wrt what would be most user friendly.

**Comment by BrunoGrandePhD** [comment]
Just to confirm we're on the same page, imagine you had the following CSV data model:

```
Attribute            ,Valid Values         ,DependsOn                  ,Required
ClinicalData         ,                     ,"PatientId,TODO"           ,FALSE
PatientId            ,                     ,                           ,TRUE
Diagnosis            ,"Cancer,Healthy"     ,                           ,TRUE
Cancer               ,                     ,"CancerType,DiagnosisDate" ,FALSE
CancerType           ,"Skin,Lung,Colon"    ,                           ,TRUE
Skin                 ,CancerClassification ,                           ,FALSE
CancerClassification ,"Benign,Malignant"   ,                           ,TRUE
DiagnosisDate        ,                     ,                           ,TRUE
Sex                  ,"Female,Male,Other"  ,                           ,TRUE

**Comment by milen-sage** [comment]
> Just to confirm we're on the same page, imagine you had the following CSV model for the `ClinicalData` data type (spaces added for readability). Currently, `schematic` would order the columns in the template manifest as follows: `Diagnosis`, `PatientId`, `Sex`, and `FavoriteColor` (_i.e._ required columns first and then alphabetical order). For simplicity, I'm ignoring the `Filename` and `entityId` special columns in this comment.
> 

The default expected order for this example should be slightly different:
PatientId,Diagnosis,Sex,FavoriteColor
i.e. required attributes for ClinicalData should follow the user-entered order in DependsOn 
@BrunoGrandePhD do you see a different order? If so which branch are you on and could you pull the latest code on for manifest/generator either from develop or main?

> ```
> Attribute            ,Valid Values         ,DependsOn                               ,Required
> ---------------------|---------------------|----------------------------------------|--------
> ClinicalData         ,                     ,"PatientId,Diagnosis,Sex,FavoriteColor" ,FALSE
> Diagnosis            ,"Cancer,Healthy"     ,                                        ,TRUE
> PatientId            ,                     ,                                        ,TRUE
> Sex                  ,"Female,Male,Other"  ,                                        ,TRUE
> FavoriteColor        ,"Red,Green,Blue"     ,                                        ,FALSE
> ```
> 
> Let's add some conditionally required attributes (drawn from the examples given in the setup docs). Let's say a `Cancer` diagnosis requires additional information, namely `CancerType` and `DiagnosisDate`, and in turn, a `Skin` cancer type would conditionally require `CancerClassification` since so many skin cancers are benign and should be considered separately. With this CSV data model (updated below), `schematic` would currently order the columns as follows: `Diagnosis`, `PatientId`, `Sex`, `CancerClassification`, `CancerType`, `DiagnosisDate`, and `FavoriteColor`.
> 

Similarly to the above the expected default order would be different:
PatientId,Diagnosis,Sex,FavoriteColor,CancerClassification,CancerType,DiagnosisDate

> ```
> Attribute            ,Valid Values         ,DependsOn                               ,Required
> ---------------------|---------------------|----------------------------------------|--------
> ClinicalData         ,                     ,"PatientId,Diagnosis,Sex,FavoriteColor" ,FALSE
> Diagnosis            ,"Cancer,Healthy"     ,                                        ,TRUE
> PatientId            ,                     ,                                        ,TRUE
> Sex                  ,"Female,Male,Other"  ,                                        ,TRUE
> FavoriteColor        ,"Red,Green,Blue"     ,                                        ,FALSE
> Cancer               ,                     ,"CancerType,DiagnosisDate"              ,FALSE
> CancerType           ,"Skin,Lung,Colon"    ,                                        ,TRUE
> DiagnosisDate        ,                     ,                                        ,TRUE
> Skin                 ,CancerClassification ,                                        ,FALSE
> CancerClassification ,"Benign,Malignant"   ,                                        ,TRUE
> ```
> 
> If I understand correctly, this issue suggests that we should interweave the conditionally required columns with the unconditionally required columns. The outstanding question is how to best determine this order. IMHO, the most intuitive order is grouping all conditionally required columns with the associated "root attribute" (_i.e._ one of the required columns for `ClinicalData`), and the conditionally required columns would be ordered based on a depth-first traversal of the graph structure.
> 

Yes, the interweaving you propose should work pretty well.

> In this case, the order would look like: `Diagnosis`, `CancerType`, `CancerClassification`, `DiagnosisDate`, `PatientId`, `Sex`, and `FavoriteColor`. Note how `CancerClassification` is listed before `DiagnosisDate` because it's conditionally required by `CancerType` depending on the latter's value. Also, I listed `CancerType` (with `CancerClassification`) before `DiagnosisDate` since ties during graph traversal would currently be resolved using alphabetical ordering. See my P.S. at the bottom of this post for a suggestion on how we can provide the user more flexibility in this regard.
> 

Note comment on order above: the "root" attributes for ClinicalData, by default, should follow the order specified by the user in the csv column "DependsOn"
 
> As far as the implementation is concerned, it seems like we would update the `ManifestGenerator.sort_manifest_fields()` method. The column order can start off as unconditionally required first (with alphabetical ordering to break ties). 

Instead of breaking the ties by alphabetical ordering, can we follow the order in the DependsOn column and then do the iteration described below for each unconditionally require column/root attribute?

>Then, we could iterate through each unconditionally required column and traverse its conditional dependency graph using a depth-first approach. Any columns encountered during this traversal would be pulled from further down in the list to being right next to the unconditionally required column in question. One aspect that I wasn't sure how to implement is differentiating between attributes that are meant to be column headers and attributes that are valid values (_e.g._ `Cancer` and `Skin` in the above example). However, as I was writing this, I realized that this was already figured out since the `sort_manifest_fields()` method is already given the list of column headers, which were presumably figured out elsewhere in the code (I'm guessing in `get_empty_manifest()`). So, as I traverse the dependency graph, I just need to check if the current node/attribute appears in the given `manifest_fields` list and if so, move that item up in the list.

Yes, that's correct, manifest_fields has the full list of header columns, those can be reordered  in sort_manifest_fields() based on the order returned by DFS described above.

> 
> **P.S.** As I looked through the sorting code, I realized that alphabetical ordering is unavoidable. 

Could you link sorting code you refer too; there are a couple of places that we need to update the schema/generator.py at least on the main branch - looks like unnecessary sorting calls have made their way here:
- https://github.com/Sage-Bionetworks/schematic/blob/41c181fcc3e56fbc879aaecc053f357f184030d2/schematic/schemas/generator.py#L82
- https://github.com/Sage-Bionetworks/schematic/blob/41c181fcc3e56fbc879aaecc053f357f184030d2/schematic/schemas/generator.py#L105
- https://github.com/Sage-Bionetworks/schematic/blob/41c181fcc3e56fbc879aaecc053f357f184030d2/schematic/schemas/generator.py#L175

the first two of those are just inefficiencies - we don't need the sort; the last one is a bug! We don't want to sort alphabetically in the return statement given we have 15 lines of code above to determine what kind of sorting is specified in the function args. We keep fixing that but I think it keeps getting committed again with PRs containing older versions of the code... We can be more careful collectively to not push old code and also catch that during reviews. I personally would try to always pull latest code before starting development on a particular feature.

>Should we consider encoding the row order in the CSV file and use that to break ties (_e.g._ how to order columns among the required vs non-required groups)? This way, the user can easily play with the ordering, such as ensuring that important columns like `PatientId` are first (after `Filename`). This could be implemented by internally creating a new column called `Priority` when loading the CSV file that contains the row number and storing this priority as a value in each node in the schema graph. This way, when traversing the graph, you will always be able to break ties by using this priority value. This solution is elegant because it doesn't involve adding another column to the CSV data model.
> 

If the code works as expected,  the user can specify the order of "root" attributes in column DependsOn; by default schema_ordered is True, so that order should be respected.

> I wonder if this approach would help simplify the following bit of code in the `SchemaGenerator.get_node_dependencies()` method because you wouldn't need to rely on the `SchemaExplorer.explore_class()` method, which is apparently slower according to the code comments. Instead, you would still use `self.get_adjacent_nodes_by_relationship()`, and the nodes would subsequently be ordered based on the priority if `schema_ordered` is `True`. I'm still learning the role for each method of the `SchemaGenerator` and `SchemaExplorer` classes, so this idea might be completely wrong.
> 
> ```
>         if schema_ordered:
>             # get dependencies in the same order in which they are defined in the schema
>             required_dependencies = self.se.explore_class(source_node)["dependencies"]
>         else:
>             required_dependencies = self.get_adjacent_nodes_by_relationship(source_node, self.requires_dependency_relationship)
> ```
> 
> If you think this priority idea is useful, I could move this text to a separate issue. I just wanted to keep it here since it's useful to have the context of the ordering problem.

Let's first confirm that when the unnecessary sorting I linked above is fixed the order of columns in a manifest by default is as expected:
- "root" attributes order follows the user specified-order in the DependOn column (btw this order is preserved in the json-ld schema generated from the csv, that's why the explore_class(source_node)["dependencies"] result is ordered accordingly)
- conditional attributes (i.e. attributes required if certain values are selected for other attributes) are ordered alphabetically
After that, applying the DFS interweaving solution above should work - using the resulting order in sort_manifest_fields(). 

Then, we can take a look at tie breaking for conditionally required attributes. We can initially have that be alphabetical, but I like a lot the idea of giving the user agency via row order.

**Comment by BrunoGrandePhD** [comment]
Sorry, I spent too much time looking at the code for `ManifestGenerator.sort_manifest_fields()` and ended up assuming that the ties are broken using alphabetical order because of the first line. 

https://github.com/Sage-Bionetworks/schematic/blob/2cd104d4f1e6a9152511233eb788a8b174313b3a/schematic/manifest/generator.py#L691

I should have actually run the code to confirm the order; I just haven't spent time to figure out how to generate the manifest template outside of the Data Curator App. My bad!

**Comment by milen-sage** [comment]
The easiest way to generate the template is manifest/examples/generator_usage.py

Just make sure the schema in the config is what you want and it contains the TEST_NODE label that you specify. E.g. ScRNA-seqLevel1 is the label of an attribute in HTAN.jsonld, so that should work for testing. You can try using example.jsonld and say Patient (for a smaller schema).

**Comment by BrunoGrandePhD** [comment]
Thanks! I think we have a plan for how to tackle this issue once the time comes.

**Comment by ychae** [comment]
@BrunoGrandePhD Is this part of your current work?

**Comment by BrunoGrandePhD** [comment]
I haven't started tackling this yet, so this might have to wait until the next phase.

**Comment by MiekoHash** [comment]
Moved to JIRA: https://sagebionetworks.jira.com/browse/FDS-212
